README.md
What is sitemap-php ?
----------

Fast and lightweight class for generating Google sitemap XML files and index of sitemap files. Written on PHP and uses XMLWriter extension (wrapper for libxml xmlWriter API) for creating XML files. XMLWriter extension is enabled by default in PHP 5 >= 5.1.2. If you having more than 50000 url, it splits items to seperated files. _(In benchmarks, 1.000.000 url was generating in 8 seconds)_

This is a slightly modified version of the original. The Sitemap class is now added to a 'SitemapPHP' namespace, and a composer document has been added.

How to use
----------

Include Sitemap.php file to your PHP document and call Sitemap class with your base domain.

	include 'src/SitemapPHP/Sitemap.php';
    
    use SitemapPHP\Sitemap;

	$sitemap = new Sitemap('http://example.com');	

Now, we need to define path for saving XML files. This can be relative like `xmls` or absolute `/path/to/your/folder` and *must be a writable folder*. In default it uses same folder with your script.

	$sitemap->setPath('xmls/');

Generated XML file names defaulted to `sitemap-*.xml`, you can customize prefix of filenames with `setFilename` method.

	$sitemap->setFilename('customsitemap');

	
We'll add sitemap url's with `addItem` method. In this method, only first parameter (location) is required.

	$sitemap->addItem('/', '1.0', 'daily', 'Today');
	$sitemap->addItem('/about', '0.8', 'monthly', 'Jun 25');
	$sitemap->addItem('/contact', '0.6', 'yearly', '14-12-2009');
	$sitemap->addItem('/otherpage');

w/ method chaining.

	$sitemap->addItem('/projects', '0.8')->addItem('/somepage')->addItem('/hiddenpage', '0.4', 'yearly', '01-01-2011')->addItem('/rss');

from a sql result, or whatever.

	$query = Doctrine_Query::create()
					->select('p.created_at, p.slug')
					->from('Posts p')
					->orderBy('p.id DESC')
					->useResultCache(true);
	$posts =  $query->fetchArray(array(), Doctrine_Core::HYDRATE_ARRAY);
    foreach ($posts as $post) {
        $sitemap->addItem('/post/' . $post['slug'], '0.6', 'weekly', $post['created_at']);
    }

If you need to change domain for sitemap instance, you can override it via `setDomain` method.

	$sitemap->setDomain('http://blog.example.com');
	
Finally we create index for sitemap files. This method also closes tags of latest generated xml file.

	$sitemap->createSitemapIndex('http://example.com/sitemap/', 'Today');
	
When you run your script, it generates and saves XML files to given path.
	
sitemap-0.xml

	<?xml version="1.0" encoding="UTF-8"?>
	<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
	 <url>
	  <loc>http://example.com/</loc>
	  <priority>1.0</priority>
	  <changefreq>daily</changefreq>
	  <lastmod>2011-04-07</lastmod>
	 </url>
	 <url>
	  <loc>http://example.com/about</loc>
	  <priority>0.8</priority>
	  <changefreq>monthly</changefreq>
	  <lastmod>2011-06-25</lastmod>
	 </url>
	 <url>
	  <loc>http://example.com/contact</loc>
	  <priority>0.6</priority>
	  <changefreq>yearly</changefreq>
	  <lastmod>2009-12-14</lastmod>
	 </url>
	 <url>
	  <loc>http://example.com/otherpage</loc>
	  <priority>0.5</priority>
	 </url>
	</urlset>
	
sitemap-index.xml

	<?xml version="1.0" encoding="UTF-8"?>
	<sitemapindex xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
	 <sitemap>
	  <loc>http://example.com/sitemap/sitemap-0.xml</loc>
	  <lastmod>2011-04-07</lastmod>
	 </sitemap>
	</sitemapindex>
	
You need to submit sitemap-index.xml to Google Sitemaps.


ChangeLog
1.2.0-stable (2015-04-17)
	* Now licensed as MIT.
	* Fixed: Fix fatal error on empty sitemap. Patch my @mkly.

1.1.0-stable (2013-12-03)
	* Changed: First sitemap is now called sitemap.xml, rather than
	  sitemap-1.xml. Patch by @mkly.

1.0.0-stable (2013-04-23)
	* First stable release.
	* Fixed: Issue #1: Writing empty sitemaps.

0.0.1-alpha (2012-01-25)
	* First release

Project was forked from: https://github.com/import/sitemap-php


SitemapPHP.php
<?php
class SitemapPHP
{
	
	private $writer;
	private $domain;
	const INDENT = 4;
	const SCHEMA = 'http://www.sitemaps.org/schemas/sitemap/0.9';
	const DEFAULT_PRIOTORY = 0.5;

	function __construct($domain) {
		$this->sendHeader();
		$this->setDomain($domain);
		$this->writer = new XMLWriter();
		$this->writer->openURI('php://output'); 
		$this->writer->startDocument('1.0', 'UTF-8'); 
		$this->writer->setIndent(self::INDENT); 
		$this->writer->startElement('urlset'); 
		$this->writer->writeAttribute('xmlns', self::SCHEMA);
	}
	
	/**
	 * Sets the root path of sitemap
	 *
	 * @param string $domain Root path of the website, starting with http://
	 * @return void
	 * @author Osman Ungur
	 */
	private function setDomain($domain)
	{
		$this->domain = $domain;
	}

	/**
	 * Gets the root path of sitemap
	 *
	 * @return string Returns the root path of sitemap
	 * @author Osman Ungur
	 */
	private function getDomain()
	{
		return $this->domain;
	}
	
	/**
	 * Send the xml header to browser
	 *
	 * @return void
	 * @author Osman Ungur
	 */
	private function sendHeader()
	{
		header("Content-type: text/xml");
	}
	
	/**
	 * Adds an item to sitemap
	 *
	 * @param string $loc URL of the page. This value must be less than 2,048 characters. 
	 * @param string $priority The priority of this URL relative to other URLs on your site. Valid values range from 0.0 to 1.0.
	 * @param string $changefreq How frequently the page is likely to change. Valid values are always, hourly, daily, weekly, monthly, yearly and never.
	 * @param string $lastmod The date of last modification of url. Unix timestamp or any English textual datetime description.. 
	 * @return void
	 * @author Osman Ungur
	 */
	public function addItem($loc, $priority = self::DEFAULT_PRIOTORY, $changefreq = NULL, $lastmod = NULL)
	{
		$this->writer->startElement('url');
		$this->writer->writeElement('loc', $this->getDomain() . $loc);
		if ($priority)    $this->writer->writeElement('priority', $priority);
		if ($changefreq)  $this->writer->writeElement('changefreq', $changefreq);
		if ($lastmod)     $this->writer->writeElement('lastmod', $this->getLastModifiedDate($lastmod));
		$this->writer->endElement();
	}

	/**
	 * Prepares given date for sitemap
	 *
	 * @param string $date Unix timestamp or any English textual datetime description
	 * @return string Year-Month-Day formatted date.
	 * @author Osman Ungur
	 */
	private function getLastModifiedDate($date)
	{
		if (!ctype_digit($date)) {
			$date = strtotime($date);
			return date('Y-m-d', $date);
		}
		else {
			return date('Y-m-d', $date);
		}
	}
	
	/**
	 * Sends the prepared sitemap to browser.
	 *
	 * @return void
	 * @author Osman Ungur
	 */
	public function render()
	{
		$this->writer->endElement();
		$this->writer->endDocument();
		$this->writer->flush();
	}
}

src/SitemapPHP/Sitemap.php
<?php

namespace SitemapPHP;

/**
 * Sitemap
 *
 * This class used for generating Google Sitemap files
 *
 * @package    Sitemap
 * @author     Osman Üngür <osmanungur@gmail.com>
 * @copyright  2009-2015 Osman Üngür
 * @copyright  2012-2015 Evert Pot (http://evertpot.com/)
 * @license    http://opensource.org/licenses/MIT MIT License
 * @link       http://github.com/evert/sitemap-php
 */
class Sitemap {

	/**
	 *
	 * @var \XMLWriter
	 */
	private $writer;
	private $domain;
	private $path;
	private $filename = 'sitemap';
	private $current_item = 0;
	private $current_sitemap = 0;

	const EXT = '.xml';
	const SCHEMA = 'http://www.sitemaps.org/schemas/sitemap/0.9';
	const DEFAULT_PRIORITY = 0.5;
	const ITEM_PER_SITEMAP = 50000;
	const SEPERATOR = '-';
	const INDEX_SUFFIX = 'index';

	/**
	 *
	 * @param string $domain
	 */
	public function __construct($domain) {
		$this->setDomain($domain);
	}

	/**
	 * Sets root path of the website, starting with http:// or https://
	 *
	 * @param string $domain
	 */
	public function setDomain($domain) {
		$this->domain = $domain;
		return $this;
	}

	/**
	 * Returns root path of the website
	 *
	 * @return string
	 */
	private function getDomain() {
		return $this->domain;
	}

	/**
	 * Returns XMLWriter object instance
	 *
	 * @return \XMLWriter
	 */
	private function getWriter() {
		return $this->writer;
	}

	/**
	 * Assigns XMLWriter object instance
	 *
	 * @param \XMLWriter $writer 
	 */
	private function setWriter(\XMLWriter $writer) {
		$this->writer = $writer;
	}

	/**
	 * Returns path of sitemaps
	 * 
	 * @return string
	 */
	private function getPath() {
		return $this->path;
	}

	/**
	 * Sets paths of sitemaps
	 * 
	 * @param string $path
	 * @return Sitemap
	 */
	public function setPath($path) {
		$this->path = $path;
		return $this;
	}

	/**
	 * Returns filename of sitemap file
	 * 
	 * @return string
	 */
	private function getFilename() {
		return $this->filename;
	}

	/**
	 * Sets filename of sitemap file
	 * 
	 * @param string $filename
	 * @return Sitemap
	 */
	public function setFilename($filename) {
		$this->filename = $filename;
		return $this;
	}

	/**
	 * Returns current item count
	 *
	 * @return int
	 */
	private function getCurrentItem() {
		return $this->current_item;
	}

	/**
	 * Increases item counter
	 * 
	 */
	private function incCurrentItem() {
		$this->current_item = $this->current_item + 1;
	}

	/**
	 * Returns current sitemap file count
	 *
	 * @return int
	 */
	private function getCurrentSitemap() {
		return $this->current_sitemap;
	}

	/**
	 * Increases sitemap file count
	 * 
	 */
	private function incCurrentSitemap() {
		$this->current_sitemap = $this->current_sitemap + 1;
	}

	/**
	 * Prepares sitemap XML document
	 * 
	 */
	private function startSitemap() {
		$this->setWriter(new \XMLWriter());
		if ($this->getCurrentSitemap()) {
			$this->getWriter()->openURI($this->getPath() . $this->getFilename() . self::SEPERATOR . $this->getCurrentSitemap() . self::EXT);
		} else {
			$this->getWriter()->openURI($this->getPath() . $this->getFilename() . self::EXT);
		}
		$this->getWriter()->startDocument('1.0', 'UTF-8');
		$this->getWriter()->setIndent(true);
		$this->getWriter()->startElement('urlset');
		$this->getWriter()->writeAttribute('xmlns', self::SCHEMA);
	}

	/**
	 * Adds an item to sitemap
	 *
	 * @param string $loc URL of the page. This value must be less than 2,048 characters. 
	 * @param string $priority The priority of this URL relative to other URLs on your site. Valid values range from 0.0 to 1.0.
	 * @param string $changefreq How frequently the page is likely to change. Valid values are always, hourly, daily, weekly, monthly, yearly and never.
	 * @param string|int $lastmod The date of last modification of url. Unix timestamp or any English textual datetime description.
	 * @return Sitemap
	 */
	public function addItem($loc, $priority = self::DEFAULT_PRIORITY, $changefreq = NULL, $lastmod = NULL) {
		if (($this->getCurrentItem() % self::ITEM_PER_SITEMAP) == 0) {
			if ($this->getWriter() instanceof \XMLWriter) {
				$this->endSitemap();
			}
			$this->startSitemap();
			$this->incCurrentSitemap();
		}
		$this->incCurrentItem();
		$this->getWriter()->startElement('url');
		$this->getWriter()->writeElement('loc', $this->getDomain() . $loc);
		$this->getWriter()->writeElement('priority', $priority);
		if ($changefreq)
			$this->getWriter()->writeElement('changefreq', $changefreq);
		if ($lastmod)
			$this->getWriter()->writeElement('lastmod', $this->getLastModifiedDate($lastmod));
		$this->getWriter()->endElement();
		return $this;
	}

	/**
	 * Prepares given date for sitemap
	 *
	 * @param string $date Unix timestamp or any English textual datetime description
	 * @return string Year-Month-Day formatted date.
	 */
	private function getLastModifiedDate($date) {
		if (ctype_digit($date)) {
			return date('Y-m-d', $date);
		} else {
			$date = strtotime($date);
			return date('Y-m-d', $date);
		}
	}

	/**
	 * Finalizes tags of sitemap XML document.
	 *
	 */
	private function endSitemap() {
		if (!$this->getWriter()) {
			$this->startSitemap();
		}
		$this->getWriter()->endElement();
		$this->getWriter()->endDocument();
	}

	/**
	 * Writes Google sitemap index for generated sitemap files
	 *
	 * @param string $loc Accessible URL path of sitemaps
	 * @param string|int $lastmod The date of last modification of sitemap. Unix timestamp or any English textual datetime description.
	 */
	public function createSitemapIndex($loc, $lastmod = 'Today') {
		$this->endSitemap();
		$indexwriter = new \XMLWriter();
		$indexwriter->openURI($this->getPath() . $this->getFilename() . self::SEPERATOR . self::INDEX_SUFFIX . self::EXT);
		$indexwriter->startDocument('1.0', 'UTF-8');
		$indexwriter->setIndent(true);
		$indexwriter->startElement('sitemapindex');
		$indexwriter->writeAttribute('xmlns', self::SCHEMA);
		for ($index = 0; $index < $this->getCurrentSitemap(); $index++) {
			$indexwriter->startElement('sitemap');
			$indexwriter->writeElement('loc', $loc . $this->getFilename() . ($index ? self::SEPERATOR . $index : '') . self::EXT);
			$indexwriter->writeElement('lastmod', $this->getLastModifiedDate($lastmod));
			$indexwriter->endElement();
		}
		$indexwriter->endElement();
		$indexwriter->endDocument();
	}

}
